apiserver/params/network.go

```diff
@@ -1226,3 +1236,111 @@ type SubnetsResult struct {
 type SubnetsResults struct {
 	Results []SubnetsResult `json:"results"`
 }
+
+// SpaceInfosParams provides the arguments for a SpaceInfos call.
+type SpaceInfosParams struct {
+	// A list of space IDs for filtering the returned set of results. If
+	// empty, all spaces will be returned.
+	FilterBySpaceIDs []string `json:"space-ids,omitempty"`
+}
+
+// SpaceInfos represents the result of a SpaceInfos API call.
+type SpaceInfos struct {
+	Infos []SpaceInfo `json:"space-infos,omitempty"`
+}
+
+// SpaceInfo describes a space and its subnets.
+type SpaceInfo struct {
+	ID         string     `json:"id"`
+	Name       string     `json:"name"`
+	ProviderID string     `json:"provider-id,omitempty"`
+	Subnets    []SubnetV3 `json:"subnets,omitempty"`
+}
+
+// FromFromNetworkSpaceInfos converts a network.SpaceInfos into a serializable
+// payload that can be sent over the wire.
+func FromNetworkSpaceInfos(allInfos network.SpaceInfos) SpaceInfos {
+	res := SpaceInfos{
+		Infos: make([]SpaceInfo, len(allInfos)),
+	}
+
+	for i, si := range allInfos {
+		mappedSubnets := make([]SubnetV3, len(si.Subnets))
+		for j, subnetInfo := range si.Subnets {
+			var mappedFanInfo *FanConfigEntry
+			if subnetInfo.FanInfo != nil {
+				mappedFanInfo = &FanConfigEntry{
+					Underlay: subnetInfo.FanInfo.FanLocalUnderlay,
+					Overlay:  subnetInfo.FanInfo.FanOverlay,
+				}
+			}
+
+			mappedSubnets[j] = SubnetV3{
+				SpaceID:  subnetInfo.SpaceID,
+				FanInfo:  mappedFanInfo,
+				IsPublic: subnetInfo.IsPublic,
+
+				SubnetV2: SubnetV2{
+					ID: string(subnetInfo.ID),
+					Subnet: Subnet{
+						CIDR:              subnetInfo.CIDR,
+						ProviderId:        string(subnetInfo.ProviderId),
+						ProviderNetworkId: string(subnetInfo.ProviderNetworkId),
+						ProviderSpaceId:   string(subnetInfo.ProviderSpaceId),
+						VLANTag:           subnetInfo.VLANTag,
+						Zones:             subnetInfo.AvailabilityZones,
+						// NOTE(achilleasa): the SpaceTag is not populated
+						// as we can grab the space name from the parent
+						// SpaceInfo when unmarshaling.
+					},
+				},
```

Comment:
I wonder if there is a cleaner structure for this. A V3 holding a V2 holding a V1 doesn't inspire me, and is sort of the inverse of how we do Facade versioning. (As we go V4, etc, it just gets worse.)
Doesn't need to change here, but something we should consider.